sympy/assumptions/assume.py

```diff
@@ -86,80 +132,226 @@ def arg(self):
+    dispatch the handler by argument types.
 
-        >>> from sympy import Q, ask
-        >>> type(Q.prime)
-        <class 'sympy.assumptions.assume.Predicate'>
-        >>> Q.prime.name
-        'prime'
-        >>> Q.prime(7)
-        Q.prime(7)
-        >>> _.func.name
-        'prime'
+    If you directly construct predicate using this class, you will get
+    ``UndefinedPredicate`` which cannot be dispatched. This is useful
+    when you are building boolean expressions which do not need to be
+    evaluated.
 
-    To obtain the truth value of an expression containing predicates, use
-    the function ``ask``:
+    Examples
+    ========
+
+    Applying and evaluating to boolean value:
 
-        >>> ask(Q.prime(7))
-        True
+    >>> from sympy import Q, ask
+    >>> from sympy.abc import x
+    >>> ask(Q.prime(7))
+    True
+
+    You can define a new predicate by subclassing and dispatching. Here,
+    we define a predicate for sexy primes [2] as an example.
+
+    >>> from sympy import Predicate, Integer
+    >>> class SexyPrimePredicate(Predicate):
+    ...     name = "sexyprime"
+    >>> Q.sexyprime = SexyPrimePredicate()
+    >>> @Q.sexyprime.register(Integer, Integer)
+    ... def _(int1, int2, assumptions):
+    ...     args = sorted([int1, int2])
+    ...     if not all(ask(Q.prime(a), assumptions) for a in args):
+    ...         return False
+    ...     return args[1] - args[0] == 6
+    >>> ask(Q.sexyprime(5, 11))
+    True
+
+    Direct constructing returns ``UndefinedPredicate``, which can be
+    applied but cannot be dispatched.
+
+    >>> from sympy import Predicate, Integer
+    >>> Q.P = Predicate("P")
+    >>> type(Q.P)
+    <class 'sympy.assumptions.assume.UndefinedPredicate'>
+    >>> Q.P(1)
+    Q.P(1)
+    >>> Q.P.register(Integer)(lambda expr, assump: True)
+    Traceback (most recent call last):
+      ...
+    TypeError: <class 'sympy.assumptions.assume.UndefinedPredicate'> cannot be dispatched.
 
     The tautological predicate ``Q.is_true`` can be used to wrap other objects:
+    >>> Q.is_true(x > 1)
+    Q.is_true(x > 1)
 
-        >>> from sympy.abc import x
-        >>> Q.is_true(x > 1)
-        Q.is_true(x > 1)
+    References
+    ==========
+
+    .. [1] https://en.wikipedia.org/wiki/Predicate_(mathematical_logic)
+    .. [2] https://en.wikipedia.org/wiki/Sexy_prime
 
     """
 
     is_Atom = True
 
+    def __new__(cls, *args, **kwargs):
+        if cls is Predicate:
+            return UndefinedPredicate(*args, **kwargs)
+        obj = super().__new__(cls, *args)
+        return obj
+
+    @property
+    def name(self):
+        # May be overridden
+        return type(self).__name__
+
+    def register(self, *types, **kwargs):
+        if self.handler is None:
+            # condition for UndefinedPredicate
+            raise TypeError("%s cannot be dispatched." % type(self))
+        return lambda func: self.handler.register(*types, **kwargs)(func)
```

Comment:
Any reason for the lambda here?
```suggestion
        return self.handler.register(*types, **kwargs)
```